compiler/llvm/clangUtil.cpp

```diff
@@ -3018,84 +3084,141 @@ void makeBinaryLLVM(void) {
-    if (err) {
-      USR_FATAL("removing file %s failed: %s\n",
-                executableFilename,
-                err.message().c_str());
-    }
-
-    // mv tmp/hello.tmp hello
-    if( printSystemCommands )
-      printf("mv %s %s\n", tmpbinname, executableFilename);
-
-    err = llvm::sys::fs::rename(tmpbinname, executableFilename);
-    if (err) {
-      // But that might fail if /tmp is on a different filesystem.
-
-      std::string mv("mv ");
-      mv += tmpbinname;
-      mv += " ";
-      mv += executableFilename;
-
-      mysystem(mv.c_str(), mv.c_str());
-
-      /* For future LLVM,
-      err = llvm::sys::fs::copy_file(tmpbinname, executableFilename);
-      if (err) {
-        USR_FATAL("copying file %s to %s failed: %s\n",
-                  tmpbinname,
-                  executableFilename,
-                  err.message().c_str());
-      }
-
-      // and then set permissions, like mv
-      auto maybePerms = llvm::sys::fs::getPermissions(tmpbinname);
-      if (maybePerms.getError()) {
-        USR_FATAL("reading permissions on %s failed: %s\n",
-                  tmpbinname,
-                  err.message().c_str());
-      }
-      err = llvm::sys::fs::setPermissions(executableFilename, *maybePerms);
-      if (err) {
-        USR_FATAL("setting permissions on %s failed: %s\n",
-                  executableFilename,
-                  err.message().c_str());
-      }
-
-      // and then remove the file, so it's like mv
-      err = llvm::sys::fs::remove(tmpbinname);
-      if (err) {
-        USR_FATAL("removing file %s failed: %s\n",
-                  tmpbinname,
-                  err.message().c_str());
-      }*/
-    }
-
-  } else {
-    // Now run the makefile to move from tmpbinname to the proper program
-    // name and to build a launcher (if necessary).
-    const char* makeflags = printSystemCommands ? "-f " : "-s -f ";
-    const char* makecmd = astr(astr(CHPL_MAKE, " "),
-                               makeflags,
-                               getIntermediateDirName(), "/Makefile");
-
-    if( printSystemCommands ) {
-      printf("%s\n", makecmd);
-      fflush(stdout); fflush(stderr);
-    }
-
-    mysystem(makecmd, "Make Binary - Building Launcher and Copying");
+static void moveGeneratedLibraryFile(const char* tmpbinname) {
+  // Need to reuse some of the stuff in codegen_makefile.  It doesn't save the
+  // full filename that is used when in library mode, so we don't have an
+  // alternative to making a modified version of executableFilename again
+  const char* exeExt = getLibraryExtension();
+  const char* libraryPrefix = "";
+  int libLength = strlen("lib");
+  bool startsWithLib = strncmp(executableFilename, "lib", libLength) == 0;
+  if (!startsWithLib) {
+    libraryPrefix = "lib";
+  }
+  const char* fullLibraryName = astr(libDir, "/", libraryPrefix,
+                                     executableFilename, exeExt);
+
+  std::error_code err;
+
+  // rm -f hello
```

Comment:
Could we have a new function in this file like `moveResultFromTmp` that does the removing / moving / copying across filesystems?